[Accessibility] Missing status message on Cart and Checkout

Manussakis opened the pull request:
… classic cart

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #51902 .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

**Remove item from Cart**
The following steps should be done to test both the classic and the block-based Cart.
- Add a product to the Cart.
- Go to the cart page.
- Turn on the screen reader.
- Navigate using the keyboard.
- Remove the item from the Cart.
- For classic Cart: Verify the notice message that confirms the item was successfully removed is announced.
- For block-based Cart: Verify the item removal is announced and there is no visual alert message.

**HEADS-UP**
Different behaviors can be seen when using a screen reader with a mouse. The goal is to ensure the functionality works as expected with the keyboard navigation because:
1. Most of the screen reader users navigate through the page using the keyboard.
2. Screen readers have different settings to determine how the mouse navigation should behave.
VoiceOver example:
![image](https://github.com/user-attachments/assets/f8126941-3350-4a58-8c73-514354ead19d)
3. Mobile devices don't have the mouse navigation option.  


**Update item quantity**
- Add a product to the block-based Cart.
- Go to the Cart page.
- Turn on the screen reader.
- Increase and decrease the item quantity.
- Verify the actions are announced by screen readers.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

Manussakis committed:
Trigger item_removed_from_classic_cart event when removing items from classic cart

Manussakis committed:
Prevent shipping found alert to be announced if it didn't change when updating items quantity

Manussakis committed:
Use createNotice instead of speak function to alert items removal from Cart

Manussakis committed:
Merge branch 'trunk' into fix/51902

Manussakis committed:
Merge branch 'trunk' into fix/51902

Manussakis committed:
Add changelog file

Manussakis committed:
Merge branch 'trunk' into fix/51902

Manussakis commented:
@dkotter This PR is ready for code review.
It's worth checking the discussion on the issue to get more context.
Thanks!

dkotter reviewed (approved):
@qasumitbagthariya @ankitguptaindia This is ready for QA

qasumitbagthariya commented:
### QA/Test Report ✅ 
-------------

I have checked this PR in the `Manussakis:fix/51902` branch and changes working as expected.


**Testing Environment** 

<Details>

- WordPress: 6.7.1
- Theme: Storefront 4.6.0
- WooCommerce - WooCommerce - Dev version with this fix specific branch.
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 
- Git Branch: Manussakis:fix/51902

</Details>


**Remove item from Cart ✅**

 - verified with both the classic and the block-based Cart.
 
https://github.com/user-attachments/assets/040fd78f-d2d6-4f0c-aa1a-19165fe351f1


https://github.com/user-attachments/assets/008f8153-ee9d-4556-a809-97bff4a14efe

**Update item quantity ✅**

https://github.com/user-attachments/assets/49c737c8-8464-428e-8432-f7b65691f3fb


@Manussakis please remove [WIP] from the title and take this PR out of draft.

vikrampm1 commented:
@mikejolley this PR is now ready for your review!

Manussakis committed:
Merge branch 'trunk' into fix/51902

mikejolley commented on plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/shipping-rates-control/index.tsx:
```diff
@@ -70,14 +70,37 @@ const ShippingRatesControl = ( {
+			shippingRatesPackageCount.current === newShippingRatesPackageCount
+		) {
+			return;
+		}
```
Is this code to prevent the same values being spoken multiple times? I think what you have is fine but it would be my preference to avoid the refs and instead use the `usePrevious` hook.

```
import { usePrevious } from '@woocommerce/base-hooks';
```

This will let you store the previous value to a variable, which you can then compare in the `useEffect` hook to ensure it only speaks when a change has occurred. What do you think?

Manussakis replied:
@mikejolley I didn't know about the `usePrevious` hook. I think using it makes sense to keep consistency with the rest of the codebase.

I have made the changes, please review it:

- [Use usePrevious instead of useRef to track changing in values](https://github.com/woocommerce/woocommerce/pull/53494/commits/9f7516e2518331aeaa1b67fd875c595960844ed2)

Thanks!

cc: @dkotter

mikejolley reviewed (commented):
Looks good but I have made a suggestion for changing the refs to a different hook. Could you take a look?

Manussakis committed:
Merge branch 'trunk' into fix/51902

Manussakis committed:
Use usePrevious instead of useRef to track changing in values

mikejolley reviewed (changes requested):
Thanks for making the change; thats much cleaner.

I was unsure about the visual snackbar notice on the cart block and after discussing with team we don't think we should include it. 

With the cart block, there is no page refresh and the visual feedback is immediate—you can see the item removed from the cart. The snackbar adds visual noise and while it can be dismissed, its not necessary.

The speak is still needed however. 

If you agree, could we revert that part and keep the other changes?

I will be going AFK today so this will need approval from someone else in my absence.

jeffpaul commented:
@Manussakis back to you for a bit of partial revert, please and thanks!

Manussakis commented:
Thanks @mikejolley !
I am moving the issue to In Progress to work on your feedback.

Manussakis committed:
Merge branch 'trunk' into fix/51902

Manussakis committed:
Remove notice from block-based cart once item is removed

Manussakis commented:
@mikejolley

I have addressed the feedback you left [in this comment](https://github.com/woocommerce/woocommerce/pull/53494#pullrequestreview-2517687462). 
Please check the [Remove notice from block-based cart once item is removed](https://github.com/woocommerce/woocommerce/pull/53494/commits/71ee5ff98799909e6e67e901f01959ef0867d7a2) commit out.

cc: @dkotter

vikrampm1 commented:
@mikejolley this is back for your re-review!

mikejolley commented on plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/shipping-rates-control/index.tsx:
```diff
@@ -70,15 +71,39 @@ const ShippingRatesControl = ( {
+	const shippingRatesPackageCount =
+		getShippingRatesPackageCount( shippingRates );
+	const previousShippingRatesRateCount = usePrevious(
+		getShippingRatesRateCount( shippingRates )
```
Would this not be more efficient to use the `shippingRatesRateCount` const already defined above?

Manussakis replied:
Somehow I overlooked that. Good catch, thanks!

Fixed in [Reuse variables as parameters](https://github.com/woocommerce/woocommerce/pull/53494/commits/93478679d1e863059b7078f8c255a6523f009eb2) commit.

mikejolley commented on plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/shipping-rates-control/index.tsx:
```diff
@@ -70,15 +71,39 @@ const ShippingRatesControl = ( {
+		getShippingRatesRateCount( shippingRates )
+	);
+	const previousShippingRatesPackageCount = usePrevious(
+		getShippingRatesPackageCount( shippingRates )
```
As above, we already have `shippingRatesPackageCount`. Would it work to reuse it?

Manussakis replied:
Fixed in [Reuse variables as parameters](https://github.com/woocommerce/woocommerce/pull/53494/commits/93478679d1e863059b7078f8c255a6523f009eb2) commint.

Manussakis committed:
Merge branch 'trunk' into fix/51902

Manussakis committed:
Reuse variables as parameters

mikejolley commented:
@Manussakis Thanks for that change. Can you confirm what should happen on classic cart? In the video above I see "Your cart is currently empty." is announced, but in my test store its not. Instead it just tells you where you are on the page after removal.

Block based is working.

Manussakis commented:
@mikejolley 

> Can you confirm what should happen on classic cart? In the video above I see "Your cart is currently empty." is announced, but in my test store its not. Instead it just tells you where you are on the page after removal.

I was not able to reproduce it either, and that's not the expected behavior.

When the last item is removed from the Classic Cart, the HTML response comes with two notices:
- [Item] removed. Undo? – `role="alert"`
- Your cart is currently empty. – `role="status"`

Both notices come as live regions, but because live regions are not announced if they come populated from the server, neither was announced before this fix.
What we did along this PR was to [call the `focus_populate_live_region` function](https://github.com/woocommerce/woocommerce/pull/53494/files#diff-d0cd5e4193aa92c6a82831c26d0b53846af5c69ec991f39f33afe9312adefb3dR185), so the "[Item] removed. Undo?" notice is focused and then voiced.

The "Your cart is currently empty." notice should never be announced either because it comes populated from the serve or because on the Storefront theme it's not a live region, since there is no `role="alert"` or `role="status"` attribute added to it.
![Screenshot 2025-01-09 at 16 20 30](https://github.com/user-attachments/assets/c1c41e3b-6234-4257-99e5-8c852e6cbf31)

This seems an issue out of the scope of this one, because it happens even when the Classic Cart page is opened without products. If you agree, I can create a new issue to fix that.

Going back to your original question:

> Can you confirm what should happen on classic cart?

Only the "[Item] removed. Undo?" notice should be voiced.

This brings us to a new point: What should happen when the last item is removed from the Classic Cart?
My opinion is that only the "Your cart is currently empty" notice should be announced. This would provide users with two pieces of information in one message:
- The item has been removed.
- The cart is now empty.

@mikejolley  Can I implement this change?

@qasumitbagthariya Could you please reproduce what happens in the [first video of the QA report](https://github.com/woocommerce/woocommerce/pull/53494#issuecomment-2545398197)?
The report says the tested theme is Storefront, but I couldn't reproduce it 🤔.

mikejolley commented:
@Manussakis I get the notice now when using keyboard to navigate. If you remove the item from cart with mouse click it doesn't announce, it just has:

![Screenshot 2025-01-10 at 11 26 18](https://github.com/user-attachments/assets/a5e491c1-7fbd-4eef-9d60-5ebb837c88c4)

I'm not sure if this is expected or not?

Manussakis commented:
@mikejolley 

> I get the notice now when using keyboard to navigate.

Do you mean the "Your cart is currently empty" notice?
If so, please let me know your environment and the steps to reproduce it.
Technically speaking, I don't understand how this notice can be announced, for the reasons described in [my previous comment](https://github.com/woocommerce/woocommerce/pull/53494#issuecomment-2581106776). Especially when using the Storefront theme.

---

> If you remove the item from cart with mouse click it doesn't announce, it just has:

I don't believe we should focus on the behavior when using a screen reader with the mouse because:
1. Most of the screen reader users navigate through the page using the keyboard.
2. Screen readers have different settings to determine how the mouse navigation should behave.
VoiceOver example:
![image](https://github.com/user-attachments/assets/f8126941-3350-4a58-8c73-514354ead19d)
3. Mobile devices don't have the mouse navigation option.

mikejolley commented:
> Do you mean the "Your cart is currently empty" notice?

No, the undo notice. It works as you described. 

> I don't believe we should focus on the behavior when using a screen reader with the mouse because:

In that case can the testing instructions be updated to show this?

mikejolley committed:
Merge branch 'trunk' into fix/51902

mikejolley commented:
Cheers, merging.
